cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+                                                     thrust::plus<uint32_t>());
+
+  auto d_quad_offset = [&]() {
+    // allocate memory for both numbers and their prefix-sums
+    rmm::device_uvector<uint32_t> d_num_sums(num_original_pairs, stream);
+    rmm::device_uvector<uint32_t> d_num_units(num_original_pairs, stream);
+
+    // computes numbers of sub-pairs for each quadrant-polygon pair
+    thrust::transform(rmm::exec_policy(stream)->on(stream),
+                      quad_indices.begin<uint32_t>(),
+                      quad_indices.end<uint32_t>(),
+                      d_num_units.begin(),
+                      get_num_units);
+
+    thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           d_num_units.begin(),
+                           d_num_units.end(),
+                           d_num_sums.begin());
+
+    // allocate memory for sub-pairs with four components:
+    // (polygon_idx, quadrant_idx, offset, length)
+
+    rmm::device_uvector<uint32_t> d_quad_offset(num_unit_pairs, stream);
+
+    thrust::fill(
+      rmm::exec_policy(stream)->on(stream), d_quad_offset.begin(), d_quad_offset.end(), 0);
+
+    // scatter 0..num_org_pair to d_quad_offset using d_num_sums as map
+    thrust::scatter(rmm::exec_policy(stream)->on(stream),
+                    thrust::make_counting_iterator(0),
+                    thrust::make_counting_iterator(0) + num_original_pairs,
+                    d_num_sums.begin(),
+                    d_quad_offset.begin());
+
+    // copy idx of orginal pairs to all sub-pairs
+    thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+                           d_quad_offset.begin(),
+                           d_quad_offset.end(),
+                           d_quad_offset.begin(),
+                           thrust::maximum<int>());
+
+    // d_num_sums and d_num_units are no longer needed, so don't return them
+
+    return std::move(d_quad_offset);
+  }();
+
+  rmm::device_uvector<uint32_t> d_pq_poly_idx(num_unit_pairs, stream);
+  rmm::device_uvector<uint32_t> d_pq_quad_idx(num_unit_pairs, stream);
+  rmm::device_uvector<uint32_t> d_quad_length(num_unit_pairs, stream);
+
+  // gather polygon idx and quadrant idx from original pairs into sub-pairs
+  thrust::gather(rmm::exec_policy(stream)->on(stream),
+                 d_quad_offset.begin(),
+                 d_quad_offset.end(),
+                 poly_indices.begin<uint32_t>(),
+                 d_pq_poly_idx.begin());
+
+  thrust::gather(rmm::exec_policy(stream)->on(stream),
+                 d_quad_offset.begin(),
+                 d_quad_offset.end(),
+                 quad_indices.begin<uint32_t>(),
+                 d_pq_quad_idx.begin());
+
+  // allocate memory to store numbers of points in polygons in all sub-pairs initialized to 0
+  rmm::device_uvector<uint32_t> d_num_hits(num_unit_pairs, stream);
+  thrust::fill(rmm::exec_policy(stream)->on(stream), d_num_hits.begin(), d_num_hits.end(), 0);
```

Comment:
cudaMemsetAsync...

Actually, quad_pip_phase1_kernel seems to initialize this array (see line 136), so I don't think you need to initialize it to zero at all.